### PR TITLE
Workflow for HDMF Dev compatibility

### PR DESCRIPTION
**Pull request by mavaylon1**
## Summary of changes
Created new workflow to check for HDMf Dev compatibility.

The workflow will clone the dev branch of HDMF and run the tests for "common".

## PR checklist for schema changes

- [x] Update the version string in `docs/source/conf.py` and `common/namespace.yaml` to the next version with the suffix "-alpha"
- [x] Add a new section in the release notes for the new version with the date "Upcoming"
- [x] Add release notes for the PR to `docs/source/hdmf_common_release_notes.rst` and/or
  `docs/source/hdmf_experimental_release_notes.rst`

### CODE REVIEW

**Comment by mavaylon1** [comment]
@rly @oruebel  I took a different route to what was on nwb-schema and similar to how we test pynwb on hdmf. The way nwb-schema validates is comparing the yaml to a json file, which we do not have. I thought it would be good enough to install hdmf and just test the "common" folder.

I do have a question :
```
name: Cancel non-latest runs
        uses: styfle/cancel-workflow-action@0.11.0
        with:
          all_but_latest: true
          access_token: ${{ github.token }}
```
I am a little confused about the github.token
Further digging, I found that GITHUB_TOKEN is an environment variable that is generated during the workflow for read and write access. (https://devopsjournal.io/blog/2022/01/03/GitHub-Tokens) 

I don't fully understand the point of this.

**Comment by mavaylon1** [comment]
Note I did not go through the check list for the PR and this is still a draft. @oruebel @rly

**Comment by mavaylon1** [comment]
The token is used to authenticate in the workflow job. Does this mean, we need to give the workflow permission to read the repository or in this case to cancel the workflow job if it is running already. 

If I as a third party person (not dev) want to cancel a workflow, I cannot because I don't have permission. Is this just giving the workflow permission to cancel to workflow if it exists as if the workflow was a third party person?

**Comment by mavaylon1** [comment]
I updated the release version for just hdmf_common with a minor release since this a update to the workflows, but this is not a change to the structure so I am not sure if this is even needed.

**Comment by rly** [comment]
This workflow will checkout the dev branch of hdmf and the specific commit of hdmf-common-schema that is stored in hdmf, and then run the tests in `tests/unit/common/` of hdmf. That does not test the current hdmf-common-schema version.

What I was looking for is a little more complicated than I initially imagined. Looking at how we do it in nwb-schema (https://github.com/NeurodataWithoutBorders/nwb-schema/blob/dev/.github/workflows/validate_schema.yml), we want to run the script `validate_hdmf_spec core -m nwb.schema.json` but replacing `nwb.schema.json` with a json schema for HDMF common. It looks like we never made that schema. But that file should be pretty much the same as https://github.com/NeurodataWithoutBorders/nwb-schema/blob/dev/nwb.schema.json except that 1) the top level metadata needs to be adjusted for HDMF common, 2) all instances of "neurodata_type" -> "data_type" and 3) rename it to hdmf-common.schema.json. I think that's it. Then because the schema lives in a directory "common" instead of "core", I think the command should be `validate_hdmf_spec common -m hdmf-common.schema.json`.

**Comment by mavaylon1** [comment]
@rly what if I cd into hdmf_common_schema, git checkout PR branch for the schema, then ran the tests

**Comment by rly** [comment]
Then when the branch is merged, you have to update the workflow again so that you check out the current main branch of the schema instead of a PR branch.

That is fine to do, but I think independently of whether the dev branch of HDMF is compatible with the current PR or main branch of the schema, it would be good to run `validate_hdmf_spec common -m hdmf-common.schema.json` just like what we have in NWB schema, to catch validation issues with the schema that may not be caught by the tests in HDMF.

**Comment by mavaylon1** [comment]
> Then when the branch is merged, you have to update the workflow again so that you check out the current main branch of the schema instead of a PR branch.
> 
> 
> 
> That is fine to do, but I think independently of whether the dev branch of HDMF is compatible with the current PR or main branch of the schema, it would be good to run `validate_hdmf_spec common -m hdmf-common.schema.json` just like what we have in NWB schema, to catch validation issues with the schema that may not be caught by the tests in HDMF.

The reason why I went with this approach was because it would return an error if the schema wasn't compatible. With going the json route, does that record every instance where there's an issue? Because on second thought my way would only catch it one at a time.

**Comment by rly** [comment]
The compatibility test is good.

With JSON validation, I think it will error out at the first error. But critically, just because the schema is compatible with the current API does not mean it is valid according to the schema language. For example, if we add a new data type to the schema, there will not be corresponding code that uses that data type in the dev branch of the API (so that test will pass), but the new data type definition might be invalid according to the rules of the language. We might then write API code for this new data type only to determine later that we wrote the schema in a way that is technically not valid.

Can you please also implement a workflow that does `validate_hdmf_spec common -m hdmf-common.schema.json`?

**Comment by mavaylon1** [comment]
@rly could you do a review?

**Comment by rly** [comment]
Schema validation workflow looks good. Thanks.

The .github/workflows/hdmf_compatibility_schema.yml workflow confuses me. Currently it just checks out HDMF and runs a subset of tests from HDMF. It doesn't use the updated schema in the master branch or a pull request branch. I think it should.

The validate schema workflow was skipped. That's odd. Do you know why that happened?

**Comment by mavaylon1** [comment]
I removed "push" as a trigger because "pull_request" should trigger on each update to the PR, i.e a commit. I unerstood the trigger push, pull_request, and pull request on forked repos. However, I see now that it actually ignores the pull request trigger on the current repo, hence the need for push.

As for the compatibility workflow, I forgot to add the checkout current branch part. I let that slip that's my bad. The idea is to make sure the changes in the schema on any PR shouldn't break HDMF and if it does then it is because an update needs to be made on HDMF

**Comment by mavaylon1** [comment]
@rly I believe I made the needed changes! I need to slow down so this can be a quick check process.

File: .github/workflows/hdmf_compatibility_schema.yml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    runs-on: ubuntu-latest
+    steps:
+      - name: Cancel non-latest runs
+        uses: styfle/cancel-workflow-action@0.11.0
+        with:
+          all_but_latest: true
+          access_token: ${{ github.token }}
+      - uses: actions/checkout@v4
+        with:
+          ref: ${{ github.ref }} # checkout branch
```
Comment (rly) [inline .github/workflows/hdmf_compatibility_schema.yml:16]: I believe this checks out the current branch of this repo in the base directory. But when you run git clone hdmf, you get the version of hdmf-common that is "saved" in the default branch of hdmf. That version will be used when you run pytest, not the version checked out at the base directory which hdmf doesn't know about.

Comment (mavaylon1) [reply .github/workflows/hdmf_compatibility_schema.yml:16]: I don't want to hardcode the branch name and then have to change it each time I run it. I do see what you mean about the order of operations. I want to see if there's a way to not have manually change the branch name each time I cd into the schema repo. If not, then is this still something we want? @rly

Comment (mavaylon1) [reply .github/workflows/hdmf_compatibility_schema.yml:16]: In theory, maybe I can pass the branch name in as an expression in the workflow. I'll try that

Comment (rly) [reply .github/workflows/hdmf_compatibility_schema.yml:16]: I think `{{ github.ref }}` is the branch name. Not sure how it performs on the main branch but I think it's the same...

Comment (mavaylon1) [reply .github/workflows/hdmf_compatibility_schema.yml:16]: Yeah that's what I found as well. I'll do that tonight!

**Review by rly — APPROVED** [review]
Looks good! Thanks. Please squash and merge.

**Pull request merged**